clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1202,6 +1251,65 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
   }
 };
 
+// A type to check the validity of passing union with accessor/sampler/stream
+// member as a kernel argument types.
+class SyclKernelUnionBodyChecker : public SyclKernelFieldHandler {
+  static constexpr const bool VisitUnionBody = true;
+  int UnionCount = 0;
+  bool IsInvalid = false;
+  DiagnosticsEngine &Diag;
+
+public:
+  SyclKernelUnionBodyChecker(Sema &S)
+      : SyclKernelFieldHandler(S), Diag(S.getASTContext().getDiagnostics()) {}
+  bool isValid() { return !IsInvalid; }
+
+  bool enterUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    ++UnionCount;
+    return true;
+  }
+
+  bool leaveUnion(const CXXRecordDecl *RD, FieldDecl *FD) {
+    --UnionCount;
+    return true;
+  }
+
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+
+  /*bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
```

Comment:
Whats going on here?

Reply:
There is mismatch happening here. Diagnostic message needs to be changed here. Same for Stream types. 

/export/iusers/smanna/myworkws/llvm/clang/lib/Sema/SemaSYCL.cpp:912:59: error: no matching function for call to ‘{anonymous}::SyclKernelUnionChecker::handleSyclAccessorType(const clang::CXXBaseSpecifier&, clang::QualType&)’
             (handlers.handleSyclAccessorType(Base, BaseTy), 0)...};
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
/export/iusers/smanna/myworkws/llvm/clang/lib/Sema/SemaSYCL.cpp:1285:8: note: candidate: virtual bool {anonymous}::SyclKernelUnionChecker::handleSyclAccessorType(clang::FieldDecl*, clang::QualType)
   bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
        ^~~~~~~~~~~~~~~~~~~~~~
/export/iusers/smanna/myworkws/llvm/clang/lib/Sema/SemaSYCL.cpp:1285:8: note:   no known conversion for argument 1 from ‘const clang::CXXBaseSpecifier’ to ‘clang::FieldDecl*’
/export/iusers/smanna/myworkws/llvm/clang/lib/Sema/SemaSYCL.cpp:911:9: error: no matching function for call to ‘std::initializer_list<int>::initializer_list(<brace-enclosed initializer list>)’
         (void)std::initializer_list<int>{
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             (handlers.handleSyclAccessorType(Base, BaseTy), 0)...};
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
I have fixed this issue. I will submit new patch with diagnostic tests.

Reply:
This issue is fixed now. Patch is updated with the fix.